shared/src/crypto/signature.rs

```diff
@@ -129,69 +131,129 @@ impl quickcheck::Arbitrary for Signature {
+    /// Verifies an aggregated BLS signature. Returns `Ok(false)` if signature verification fails
+    /// and `String` error if arguments are invalid.
+    pub fn verify_bls_aggregate(
+        aggregate_sig: &[u8; BLS_SIG_LEN],
+        pub_keys: &[[u8; BLS_PUB_LEN]],
+        digests: &[[u8; BLS_DIGEST_LEN]],
+    ) -> Result<bool, String> {
+        // If the number of public keys and data does not match, return false;
+        let (num_pub_keys, num_digests) = (pub_keys.len(), digests.len());
+        if num_pub_keys != num_digests {
             return Err(format!(
-                "cannot validate a BLS signature against a {} address",
-                addr.protocol()
+                "unequal numbers of public keys ({num_pub_keys}) and digests ({num_digests})",
             ));
         }
+        if num_pub_keys == 0 {
+            return Ok(true);
+        }
 
-        let pub_k = addr.payload_bytes();
-
-        // generate public key object from bytes
-        let pk = BlsPubKey::from_bytes(&pub_k).map_err(|e| e.to_string())?;
+        // Deserialize signature bytes into a curve point.
+        let sig = BlsSignature::from_bytes(aggregate_sig)
+            .map_err(|_| "bls aggregate signature bytes are invalid G2 curve point".to_string())?;
 
-        // generate signature struct from bytes
-        let sig = BlsSignature::from_bytes(signature).map_err(|e| e.to_string())?;
+        // Deserialize each public key's bytes into a curve point.
+        let pub_keys = pub_keys
+            .iter()
+            .map(|pub_key| BlsPubKey::from_bytes(pub_key.as_slice()))
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(|_| "bls public key bytes are invalid G2 curve point".to_string())?;
+
+        // Deserialize each digest's bytes into a curve point.
+        //
+        // BLS digests and signatures are each a G2 point. The `bls_signatures` crate does not
+        // expose functionality for deserializing digest bytes into a G2 point, however it does
+        // expose bytes-to-G2 deserialization for its signature type `Signature`.
+        let digests = {
+            use bls_signatures::Signature as Digest;
+            digests
+                .iter()
+                .map(|digest| Digest::from_bytes(digest.as_slice()).map(Into::into))
+                .collect::<Result<Vec<_>, _>>()
+                .map_err(|_| "bls digest bytes are invalid G2 curve point".to_string())?
+        };
 
-        // BLS verify hash against key
-        if verify_messages(&sig, &[data], &[pk]) {
-            Ok(())
-        } else {
-            Err(format!(
-                "bls signature verification failed for addr: {}",
-                addr
-            ))
-        }
+        Ok(bls_signatures::verify(&sig, &digests, &pub_keys))
     }
 
     /// Returns `String` error if a secp256k1 signature is invalid.
     pub fn verify_secp256k1_sig(
         signature: &[u8],
-        data: &[u8],
+        hash: &[u8],
```

Comment:
see above

Reply:
NOTE: we can always ass a `verify_secp256k1_sig_digest` function (if we need it), but I'd prefer to keep this one as-is as any downstream users won't notice that we've broken it.